e2e/deploy/deploy.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deploy
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/GoogleCloudPlatform/prometheus-engine/e2e/kube"
+	"github.com/GoogleCloudPlatform/prometheus-engine/manifests"
+	"github.com/GoogleCloudPlatform/prometheus-engine/pkg/operator"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	crdsManifestFile     = "setup.yaml"
+	operatorManifestFile = "operator.yaml"
+)
+
+func CreateResources(ctx context.Context, kubeClient client.Client, deployOpts ...DeployOption) error {
+	opts := &deployOptions{}
+	for _, opt := range deployOpts {
+		opt.applyToDeployOptions(opts)
+	}
+	opts.setDefaults()
+
+	return createResources(ctx, kubeClient, func(obj client.Object) (client.Object, error) {
+		switch obj := obj.(type) {
+		case *appsv1.DaemonSet:
+			if !opts.disableGCM {
+				break
+			}
+			if obj.GetName() == operator.NameCollector {
+				for i := range obj.Spec.Template.Spec.Containers {
+					container := &obj.Spec.Template.Spec.Containers[i]
+					if container.Name == operator.CollectorPrometheusContainerName {
+						container.Args = append(container.Args, "--export.debug.disable-auth")
+						break
+					}
+				}
+			}
+		case *appsv1.Deployment:
+			if obj.GetName() == operator.NameOperator {
+				container, err := kube.DeploymentContainer(obj, "operator")
+				if err != nil {
+					return nil, fmt.Errorf("unable to find operator container: %w", err)
+				}
+				container.Args = append(container.Args, fmt.Sprintf("--project-id=%s", opts.projectID))
+				container.Args = append(container.Args, fmt.Sprintf("--location=%s", opts.location))
+				container.Args = append(container.Args, fmt.Sprintf("--cluster=%s", opts.cluster))
+				container.Args = append(container.Args, fmt.Sprintf("--operator-namespace=%s", opts.operatorNamespace))
+				container.Args = append(container.Args, fmt.Sprintf("--public-namespace=%s", opts.publicNamespace))
+			}
+		}
+		return obj, nil
+	})
+}
+
+type DeployOption interface {
+	applyToDeployOptions(*deployOptions)
```

Comment:
Question: Is there a reason to make these funcs bound to a type instead of `type DeployOption func(*deployOptions)`?